### ISSUE TITLE
LowRankArray technically incorrect

### ISSUE THREAD

**Issue by goretkin** [daanhb/Julia-tutorial#3]
I think it's a good example, but the definition of `rank` and `+` together are incorrect. For example

```
julia> A = LowRankMatrix(rand(10,2), rand(10,2));

julia> rank(A+A+A+A+A+A)
12
```

a 10x10 matrix cannot be rank 12.
You could either eagerly compress the rank when you add two `LowRankMatrix`s, or defer this to when you need to compute the rank.

**Comment by daanhb** [comment]
Glad you liked the example - and you are of course correct about its incorrectness. I went for a lazy  implementation, and perhaps the laziest solution here would be to remove the `rank` function altogether. But the example would be more compelling if things were done correctly, while still being short. That sounds more like Julia. Would rank compression be feasible in two-three lines? Perhaps by also implementing an `svd` function? I'll give it a try.

**Comment by MikaelSlevinsky** [comment]
Following [this paper](https://people.math.ethz.ch/~hiptmair/Seminars/HMAT/ST04/Papers/BGH03.pdf), I added formatted operations in SingularIntegralEquations.jl which compress the rank upon addition or subtraction (see [here](https://github.com/ApproxFun/SingularIntegralEquations.jl/blob/development/src/LinearAlgebra/FormattedOperations.jl#L31)). Though maybe the rank function should still be bounded by the minimum of the number of rows and columns.

**Comment by daanhb** [comment]
Thanks for the references! I have updated the example by adding `svd` and `recompress` functions based on these. Fortunately, it is still fairly short.

One caveat though: if you try `rank(A+A+A+A+A+A)` for a matrix of rank 2, with the first definition of `+` it gives 12. If you try it again after the corrected definition, it still gives 12. That is because the generic method for the addition of a variable number of arguments `+(a,b,c,xs,...)` is not recompiled after `+` has changed. I did not want to go there, so I've just written `rank(A+A)` to illustrate the effect of rank compression.
